### DISCUSSION TITLE
Why use TychoRepositoryTransportCacheManager?

### DISCUSSION THREAD

**Discussion by lppedd** [eclipse-tycho/tycho#2045] [General]
Why have this piece of code, for example:

https://github.com/eclipse-tycho/tycho/blob/df4dab71f344374316d22989e5eafed7aad066f0/p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/TychoRepositoryTransportCacheManager.java#L45-L53

When

https://github.com/eclipse-tycho/tycho/blob/df4dab71f344374316d22989e5eafed7aad066f0/p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/TychoRepositoryTransportCacheManager.java#L62

is going to call `handler.getLastModified` and `handler.getFile` at some point?

https://github.com/eclipse-tycho/tycho/blob/3db882dfcd422363cc4d65f553060733412425cf/p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/TychoRepositoryTransport.java#L139-L148

It would seem like we have another cache implementations for nothing. But maybe I'm missing something.  
Because of this, I have to implement another cache for FTP while it seems there is already one in the background.

**Comment by laeubi** [comment]
The cache is part of P2 (see [CacheManager](https://github.com/eclipse-equinox/p2/blob/master/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/internal/p2/repository/CacheManager.java)) what might call the Tycho transport to cache unknown transport types.

How caching works is highly depending on the protocol, e.g. HTTP works completely different than FTP, so as long as you only test with Http the super will never be called. In all other cases it calls back to the transport and probably download the file.

**Reply by lppedd** [reply]
> How caching works is highly depending on the protocol

Mmh, why would it depend on the protocol? Aren't we only interested in the last modification timestamp for the file?
Currently, the `handler.getLastModified` seems to be never called, so I don't understand why it's there.

If we keep only the overridden `getCacheDirectory` in `TychoRepositoryTransportCacheManager`, I can see that the `TransportProtocolHandler` are called correctly, whatever transport (ftp/http/https, etc.).

**Reply by laeubi** [reply]
> Mmh, why would it depend on the protocol? Aren't we only interested in the last modification timestamp for the file?

The last modification of the file is when it was last downloaded, but the question here is if the local file is out of date, you can only answer this based on some protocol specific measures (e.g. Http Expires) or at the worst case have to query the remote server and ask if it was modified in the meantime.

> If we keep only the overridden getCacheDirectory in TychoRepositoryTransportCacheManager, I can see that the TransportProtocolHandler are called correctly, whatever transport (ftp/http/https, etc.).

The default implementation is simply dumb and generates useless requests in case we know the handler already its much more performant, the cache directy p2 uses is also completely differently organized so this is of no use for Tycho implementation.

**Reply by lppedd** [reply]
@laeubi 

> The default implementation is simply dumb and generates useless requests

I suppose you're referring to the default `CacheManager`.
When I check CacheManager's `createCacheFromFile` and `createCache` methods, I see that at some point they call `TychoRepositoryTransport`, so per-se `CacheManager` doesn't do any request, it delegates to our custom `Transport`.

Inside `TychoRepositoryTransport`, the right `TransportProtocolHandler` is selected.  
So the flow seems to be correct.

If you say CacheManager's default "stale" evaluation `stale = lastModifiedRemote != lastModified;` is not optimal in some cases (e.g., for the HTTP protocol), I would suggest enhancing `TransportProtocolHandler` with a `isOptimized` method, so that `CacheManager` is not called at all.

But for other simpler protocol like FTP, the default behavior might be ok.

**Reply by laeubi** [reply]
> When I check CacheManager's createCacheFromFile and createCache methods, I see that at some point they call TychoRepositoryTransport, so per-se CacheManager doesn't do any request, it delegates to our custom Transport.

Yes ... at some point ... but we already know before any data and CacheManger would then copy the data again to a different location, that's not what we want in most of the cases.

`CacheManager` is just legacy P2 stuff we need to implement, it does not do any useful caching for Tycho use-cases.

> But for other simpler protocol like FTP, the default behavior might be ok.

FTP ist not simpler, and if you like you can just use the default, it will just be not very performant as it uses a separate cache that is most likely different from one you can use when you are doing protocol specific handling.

For example in FTP you can list all files (and there modification times) for a whole folder, there is no need to query each file separately and it would be very devastating regarding performance.

**Reply by lppedd** [reply]
@laeubi 

> and if you like you can just use the default...

The point is, you cannot really use the default. Once you register a `TransportProtocolHandler` for FTP, you're tied to implement a custom cache. No more `CacheManager` basic stuff in the current situation.

> For example in FTP you can list all files (and there modification times)...

Yeah you can do that even using the default `CacheManager` flow if you want.

> CacheManager is just legacy P2

If this legacy p2 class is going to be removed, then I understand the point in having customized cache implementations for every protocol.

**Reply by laeubi** [reply]
> No more CacheManager basic stuff in the current situation.

As said `CacheManager` is not doing anything it does not really "cache" things (even though its named that way) it just caches some very specific filetypes ... so just forget about that, you will need an own "cache" that can be just a folder where you drop a file with a structure that best fits your protocol. So a protocol must implement basically two things:

1. `lastModified` - that's nothing you can do generic
2. `fetch an URI` (that's also protocol specific) and in that process you *might* decide that you can use a previously saved file, but this must not depend on `lastModified`, as said http has many other things that are used to tell a cache "don't bother be me before ... has expired", but this is nothing FTP offers.

> If this legacy p2 class is going to be removed,

We won't convince P2 to do so, but some internal code is using it so we need to get away from it as far as we can (that's the part we implement) and the super is hopefully never called ;-)

**Reply by lppedd** [reply]
> you cannot really use the default

Let's say there is no handler registered for FTP.  
`connection.getLastModified()` will return 0, forcing a download again every time.

https://github.com/eclipse-tycho/tycho/blob/3db882dfcd422363cc4d65f553060733412425cf/p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/TychoRepositoryTransport.java#L185-L192

**Reply by lppedd** [reply]
> As said CacheManager is not doing anything it does not really "cache"

Yes, because it delegates to `TychoRepositoryTransport#getLastModified` and `TychoRepositoryTransport#download`.

But in essence it does maintain a cache

```java
if (!stale)
   return cacheFile;

// The cache is stale or missing, so we need to update it from the remote location
updateCache(cacheFile, remoteFile, lastModifiedRemote, submonitor);
return cacheFile;
```

> lastModified - that's nothing you can do generic

This one is called only if the `CacheManager` "super" flow is followed, so if you say we don't need `CacheManager`, this method can be removed.

I can even implement a new cache for FTP, no problem, it just seem entirely unnecessary to me.
